src/python_testing/TC_CLDIM_4_1.py

```diff
@@ -0,0 +1,434 @@
+    async def test_TC_CLDIM_4_1(self):
+        endpoint = self.get_endpoint(default=1)
+        timeout = self.matter_test_config.timeout if self.matter_test_config.timeout is not None else self.default_timeout
+
+        # STEP 1: Commission DUT to TH (can be skipped if done in a preceding test)
+        self.step(1)
+        attributes = Clusters.ClosureDimension.Attributes
+
+        # Default values
+        min_position = 0
+        max_position = 10000
+
+        # STEP 2a: Read FeatureMap attribute
+        self.step("2a")
+        feature_map = await self.read_cldim_attribute_expect_success(endpoint=endpoint, attribute=attributes.FeatureMap)
+
+        is_positioning_supported = feature_map & Clusters.ClosureDimension.Bitmaps.Feature.kPositioning
+        is_latching_supported = feature_map & Clusters.ClosureDimension.Bitmaps.Feature.kMotionLatching
+        is_limitation_supported = feature_map & Clusters.ClosureDimension.Bitmaps.Feature.kLimitation
+        is_speed_supported = feature_map & Clusters.ClosureDimension.Bitmaps.Feature.kSpeed
+
+        # STEP 2b: If Positioning feature is not supported, skip remaining steps
+        self.step("2b")
+        if not is_positioning_supported:
+            logging.info("Positioning feature is not supported. Skipping remaining steps.")
+            self.skip_all_remaining_steps("2c")
+            return
+
+        # STEP 2c: Read StepValue attribute
+        self.step("2c")
+        step_value = 1  # Default step value
+        step_value = await self.read_cldim_attribute_expect_success(endpoint=endpoint, attribute=attributes.StepValue)
+
+        # STEP 2d: Read LimitRange attribute
+        self.step("2d")
+        if is_limitation_supported:
+            limit_range = await self.read_cldim_attribute_expect_success(endpoint=endpoint, attribute=attributes.LimitRange)
+            min_position = limit_range.min
+            max_position = limit_range.max
+
+        # STEP 2e: Establish wildcard subscription to all attributes"
+        self.step("2e")
+        sub_handler = ClusterAttributeChangeAccumulator(Clusters.ClosureDimension)
+        await sub_handler.start(self.default_controller, self.dut_node_id, endpoint=endpoint, min_interval_sec=0, max_interval_sec=30, keepSubscriptions=False)
+
+        # STEP 2f: Read CurrentState attribute
+        self.step("2f")
+        initial_state = await self.read_cldim_attribute_expect_success(endpoint=endpoint, attribute=attributes.CurrentState)
+
+        # STEP 2g: If Latching feature is not supported or state is unlatched, skip steps 2h to 2m
+        self.step("2g")
+        if (not is_latching_supported) or (not initial_state.latch):
+            logging.info("Latching feature is not supported or state is unlatched. Skipping steps 2h to 2m.")
+            self.skip_step("2h")
+            self.skip_step("2i")
+            self.skip_step("2j")
+            self.skip_step("2k")
+            self.skip_step("2l")
+            self.skip_step("2m")
```

Comment:
comment can be applied in multiple locations
```suggestion
self.mark_step_range_skipped("2h", "2m")
```